exonum-java-binding-common/src/main/java/com/exonum/binding/common/proofs/list/ListProofStructureValidator.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.common.proofs.list;
+
+import com.exonum.binding.common.proofs.common.ProofStatus;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A validator that checks list proofs internal structure.
+ */
+public final class ListProofStructureValidator implements ListProofVisitor {
```

Comment:
Review: The fact that this validator requires two operations in client code to check the proof
seems error-prone (and there is evidence of that in the patches).
At least, I'd document how it is supposed to be used:
  - A new instance must be instantiated for each proof we check
  - accept
  - check
  - getStatus …

At most, I'd make the *constructor*
perform all the needed checks, so that the client code can look like
```
ListProofNode proof = ...
Validator v = new ListProofStructureValidator(proof);
if (v.isValid()) {

}
```
instead of
```
ListProofNode proof = ...
Validator v = new ListProofStructureValidator(proof);
v.accept(proof);
if (v.check(proof)) { // getStatus is not allowed here, you are out of luck if no check is invoked!

}
```